doc/design/python-api-1.md

```diff
@@ -0,0 +1,194 @@
+
+    * The upload & download methods (or at least the ones for uploading/downloading single assets) come in pairs:
+        * The basic methods simply upload/download everything, blocking until completion, and return either nothing or a summary of everything that was uploaded/downloaded
+            * These methods have `show_progress=True` options for whether to display progress output using pyout or to remain silent
+            * The upload methods return an `Asset` or collection of `Asset`s.  This can be implemented by having the final value yielded by the `iter_` upload function contain an `"asset"` field.
+        * Each method also has an iterator variant (named with an "`iter_`" preffix) that can be used to iterate over values containing progress information compatible with pyout
+            * These methods do not output anything (aside perhaps from logging)
+
+    * An `UploadProgressDict` is a `dict` containing some number of the following keys:
+        * `dandiset_id: str` — always present
+        * `version_id: str` — always present
+        * `path: str` — always present
+        * `status: str`
+        * `pct: float` — percentage of file digested/uploaded
+        * `current: int` — number of bytes uploaded
+        * `message: str`
+        * `size: int` — the total size of the file
+        * `errors: int`
+        * `asset: RemoteAsset` — yielded in the last `dict` of a successful upload
+    * A `DownloadProgressDict` is a `dict` containing some number of the following keys:
+        * `dandiset_id: str` — always present
+        * `version_id: str` — always present
+        * `path: str` — always present
+        * `status: str`
+        * `message: str`
+        * `size: int` — the total size of the file
+        * `current: int` — number of bytes downloaded (formerly "done")
+        * `pct: float` — percentage of bytes downloaded (formerly "done%")
+        * `checksum: str` — `"differs"`, `"ok"`, or `"-"`
+
+    * On success, blocking upload methods for uploading multiple assets return structures with the following fields:
+        * `assets: List[RemoteAsset]` — list of assets successfully uploaded
+        * `skipped: List[Tuple[LocalAsset, str]]` — list of skipped assets, paired with the reasons why they were skipped
+    * For blocking upload methods that upload multiple assets, if an error occurs while uploading, an exception is raised after all threads have completed; this exception has the same `assets` and `skipped` attributes as above plus an `errored: List[Tuple[LocalAsset, str]]` attribute pairing errored assets with error messages
+
+    * **To discuss:** Should methods that accept `DandisetMeta` and `(Bare|Remote)AssetMeta` instances also accept raw `dict`s?  If so, should any validation be done on these `dict`s?
```

Comment:
I think for starters we can just add a dedicated `@classmethod` `from_dict(cls, d: dict, validate:bool=True) -> cls (possible?)` and require such conversion explicitly.  If we keep running into needing that too often -- we would rethink

Reply:
Pydantic already provides `parse_obj()` (validating) and `construct()` (non-validating) methods for that.